src/osd/PrimaryLogPG.cc

```diff
@@ -8602,6 +8552,67 @@ void PrimaryLogPG::finish_ctx(OpContext *ctx, int log_op_type, int result)
 	   << dendl;
   utime_t now = ceph_clock_now();
 
+  // Drop the reference if deduped chunk is modified
+  if (ctx->new_obs.oi.is_dirty() &&
+    (ctx->obs->oi.has_manifest() && ctx->obs->oi.manifest.is_chunked()) &&
+    // If a clone is creating, ignore dropping the reference for manifest object
+    !ctx->delta_stats.num_object_clones &&
+    ctx->new_obs.oi.size != 0 && // missing, redirect and delete
+    log_op_type != pg_log_entry_t::PROMOTE) {
+    object_manifest_t updated_chunks;
+    object_info_t &oi = ctx->obc->obs.oi;
+    updated_chunks.type = object_manifest_t::TYPE_CHUNKED;
+    // try to find out intersections
+    for (auto &p : oi.manifest.chunk_map) {
+      if (is_dedup_chunk(oi, p.second)) {
+	if (!ctx->clean_regions.is_clean_region(p.first, p.second.length) &&
+	    p.second.oid != hobject_t()) {
+	  chunk_info_t updated_chunk;
+	  updated_chunk.offset = p.second.offset;
+	  updated_chunk.length = p.second.length;
+	  updated_chunk.oid = p.second.oid;
+	  updated_chunks.chunk_map[p.first] = updated_chunk;
+	  updated_chunks.chunk_map[p.first].set_flag(chunk_info_t::FLAG_HAS_FINGERPRINT);
```

Comment:
The chunk cannot be missing here, right?  Can we simply drop the reference?  I don't really see the point of retaining the partial dedup, it potentially wastes space over time.

Reply:
I added  `p.second.oid.oid = object_t();` to address your comment.

Reply:
I don't understand how this addresses my comment.  What does it mean for the chunk_map to have an element with an empty oid?

Reply:
You mean dropping the reference directly at here via something like dec_refcount()? Since ongoing operations are still in progress, does the reference need to be dropped after on_commit? 

> What does it mean for the chunk_map to have an element with an empty oid?

Having an empty oid means that the chunk no longer has any reference.

Reply:
That's not normally true, right?  Once we've released a reference to a chunk_map chunk we don't just clear the oid field, we remove the entry entirely.  Moreover, setting oid.oid to object_t() wouldn't be right in that case, you'd have to set oid to hobject_t() I think.  Do you mean that the code lower down uses updated_chunks to decide which references to drop later on?

Reply:
Right, to clear the chunk, oid needs to be set to hoject_t(). But, look at the following cases


 ``` 
Case A
1. set-chunk to oid A with offset 2~3
2. write oid A with offset 2~3 (content: aaa)
3. write oid A with offset 2~3 (content: bbb)
4. where the content bbb to go?
```
If we set oid to hobject_t() between 2 and 3, we don't know where the deduped chunk will be located due to lack of pool info etc. in hobject_t (we only know fingerprint oid from the content). 

To resolve this, I think there are two ways.
First, remove <offset, chunk_info> entirely and reset set-chunk. For example,
 ``` 
1. set-chunk to oid A with offset 2~3
2. write oid A with offset 2~3 (content: aaa)
2.5 clear <offset, chunk_info> in chunk_map
3. set-chunk to oid A with offset 2~3
4. write oid A with offset 2~3 (content: bbb)
 ``` 
Second is just setting oid.oid to objec_t().

What do you think?

Reply:
What oid.oid are you setting here?  In 2.5 above you removed it from chunk_map entirely, right?  Also, isn't the rest of your example:

1. set-chunk to oid A with offset 2~3
2. write oid A with offset 2~3 (content: aaa)
2.5 clear <offset, chunk_info> in chunk_map
3. set-chunk to oid A with offset 2~3
4. write oid A with offset 2~3 (content: bbb)
4.5 clear <offset, chunk_info> in chunk_map

where 2.5 and 4.5 occur in the same transaciton as 2 and 4 respectively?

Reply:
> where 2.5 and 4.5 occur in the same transactions as 2 and 4 respectively?

See below example, no need to reset set-chunk
 ``` 
1. set-chunk to oid A with offset 2~3
2. write oid A with offset 2~3 (content: aaa)
2.5 set oid.oid to object_t()
3. write oid A with offset 2~3 (content: bbb)
3.5 set oid.oid to object_t()
```

> In 2.5 above you removed it from chunk_map entirely, right? 
Right.

Reply:
I'm confused.

You changed the statement in 2.5 from "clear <offset, chunk_info> in chunk_map" to "set oid.oid to object_t()".  These two things I think are different.

The set-chunk in 3. in my example was borrowed from your second example above.  Without it, there'd be no reason for your 3.5 to exist -- we already overwrote the chunk and removed the reference.

At a higher level, I do not understand what the purpose of retaining an entry in chunk_map with an empty oid would be.  Is there some other state we need to remember about that chunk?  How would such a chunk be different from one we've *never* had a reference to?

Reply:
More specific example (set object_t()) is
``` 
1. set-chunk to oid A with offset 2~3
2. write oid A with offset 2~3 (content: aaa)
3. Generate fingerprint id (FID) from aaa
4. write oid A with offset 2~3 (content: bbb)
4.5 set oid.oid to object_t() and drop the reference of FID 
5. Generate fingerprint id from bbb (can generate hobject_t with pool location 
because we do not touch pool info in hboject_t)
```
In this case, using set-chunk means that configured offset range always is deduped even if modification happens.


And, reset set-chunk case
``` 
1. set-chunk to oid A with offset 2~3
2. write oid A with offset 2~3 (content: aaa)
3. Generate fingerprint id (FID) from aaa
4. set-chunk to oid A with offset 2~3
5. write oid A with offset 2~3 (content: bbb)
5.5 clear <offset, chunk_info> in chunk_map and drop the reference of FID
6. Generate fingerprint id from bbb (can generate hobject_t by the set-chunk in 4)
``` 
In this case, using set-chunk means that configured offset range will be deduped only once if modification occur.

I thinks both cases as above are fine. The only thing we need to discuss is set-chunk usage, which means that
"if chunk is overwritten, set-chunk should be set again ?" or not to be set again ?


> At a higher level, I do not understand what the purpose of retaining an entry in chunk_map with an empty oid would be. Is there some other state we need to remember about that chunk? How would such a chunk be different from one we've never had a reference to?
  
So, do you think clearing <offset, chunk_info> and resetting set-chunk is more Intuitive and right way?
If so, that is fine to me. I just confused what design for which we intend

Reply:
> More specific example (set object_t()) is
> 
> ```
> 1. set-chunk to oid A with offset 2~3
> 2. write oid A with offset 2~3 (content: aaa)
> 3. Generate fingerprint id (FID) from aaa

I think you are missing a 2.5 here.  Is 3 supposed to be a separate operation, or part of the same transaction as 2.?

> 4. write oid A with offset 2~3 (content: bbb)
> 4.5 set oid.oid to object_t() and drop the reference of FID 
> 5. Generate fingerprint id from bbb (can generate hobject_t with pool location 
> because we do not touch pool info in hboject_t)
> ```
> 
> In this case, using set-chunk means that configured offset range always is deduped even if modification happens.

Nothing I've seen in your your set_chunk rados operation promises that the extent will always be dedup'd in the future -- as far as I understand it purely redirects a sub-extent of the object to a target object.  I also think that such a promise would be quite unwise because:
1. a write to that extent indicates that the extent may be hot and likely to see writes in the near future
2. the new contents may not match other extents (may not dedup well)
3. the fingerprinting algorithm may produce different boundaries anyway

Once a dedup'd chunk is mutated, I claim that we should simply forget that that chunk was ever dedup'd and let the same process that decided to dedup it originally eventually decide to dedup it again if it proves to overlap with other extents and be cold enough.

> > At a higher level, I do not understand what the purpose of retaining an entry in chunk_map with an empty oid would be. Is there some other state we need to remember about that chunk? How would such a chunk be different from one we've never had a reference to?
> 
> So, do you think clearing <offset, chunk_info> and resetting set-chunk is more Intuitive and right way?
> If so, that is fine to me. I just confused what design for which we intend

I think that each entry in chunk_map costs us space in the object_info where it is encoded.  That space is *very expensive*.  As far as I can tell, your purpose in retaining an empty chunk mapping is simply that we want to know that that extent had been dedup'd in the past so that we can re-dedup it in the future on the same boundaries.
1. Is that actually your own rationale?
2. Is that really valuable?  Keep in mind, what I mentioned above: the extent is more likely now to see future writes, the write may well have changed the fingerprint boundaries, and the new contents may not duplicate other extents in the cluster.

Unless I'm missing something, retaining a memory of an overwritten chunk boundary by retaining a chunk_map entry with an empty reference is both wasteful and excessively complicated.

Reply:
I agree on the space overhead. This is definitely expensive. Apart from this, I have a concern about the space overhead of the reference (hobject_t) which is stored in the chunk object (This is not just count). I probably should solve this overhead in the near future. 

Anyway, since our design is to find out the optimal offset and length for dedup prior to do dedup, it is more clear to re-set set-chunk. Any other comments?

>I think you are missing a 2.5 here. Is 3 supposed to be a separate operation, or part of the same transaction as 2.?

a separate operation

> a write to that extent indicates that the extent may be hot and likely to see writes in the near future
the new contents may not match other extents (may not dedup well)
the fingerprinting algorithm may produce different boundaries anyway

Ok.